File: scala-server/app/com/bolour/eagerwords/server/domain/PieceProvider.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.server.domain

import com.bolour.eagerwords.common.domain.{Piece, PieceProviderType}
import com.bolour.eagerwords.common.domain.PieceProviderType.PieceProviderType
import com.bolour.util.FrequencyDistribution

import scala.util.{Success, Try}

/**
  * Factory for generating new pieces as pieces are consumed in plays.
  *
  * Note. Piece provider properties are transitive properties of games,
  * and need to be saved to and stored in the database. Hence no side effects.
  */
sealed abstract class PieceProvider {

  def take(): Try[(PieceProvider, Piece)]

  def takePieces(num: Int): Try[(PieceProvider, List[Piece])] = takePiecesAux(List(), num)

  def takePiecesAux(pieces: List[Piece], n: Int): Try[(PieceProvider, List[Piece])] = {
    if (n == 0)
      return Success((this, pieces))
    for {
      (provider1, piece) <- take()
      (provider2, pieces) <- provider1.takePiecesAux(piece +: pieces, n - 1)
    } yield (provider2, pieces)
  }

  def swapOne(piece: Piece): Try[(PieceProvider, Piece)] = {
    for {
      (provider1, resultPiece) <- this.take()
    } yield (provider1, resultPiece)
  }
}

/**
  * Cyclic piece generator - cycles through a given list of letters generating successive pieces.
  *
  * @param valueList The list of letters to cycle through.
  */
case class CyclicPieceProvider(valueList: String, counter: BigInt) extends PieceProvider {
  if (valueList == null || valueList.isEmpty)
    throw new IllegalArgumentException("empty cyclic generator value list")

  override def take(): Try[(PieceProvider, Piece)] = Try {
    val cycleLength: Int = valueList.length
    val index: Int = (counter % cycleLength).toInt
    val value = valueList(index)
    val id = counter.toString
    val piece = Piece(value, id)
    val nextPieceProvider = CyclicPieceProvider(valueList, counter + 1)
    (nextPieceProvider, piece)
  }
}

//object CyclicPieceProvider {
//  def apply() = new CyclicPieceProvider("ABCDEFGHIJKLMNOPQRSTUVWXYZ", 0)
//}

/**
  * Random piece generator - generate pieces with random letters.
  */
case class RandomPieceProvider(letterDistribution: FrequencyDistribution[Char], counter: BigInt) extends PieceProvider {

  def take(): Try[(RandomPieceProvider, Piece)] = Try {
    val letter = letterDistribution.randomValue()
    val id = counter.toString
    val piece = Piece(letter, id)
    val nextPieceProvider = RandomPieceProvider(letterDistribution, counter + 1)
    (nextPieceProvider, piece)
  }
}

//object RandomPieceProvider {
//  def apply(letterDistribution: FrequencyDistribution[Char]) = new RandomPieceProvider(letterDistribution, 0)
//}

File: haskell-server/src/EagerWords/Server/Domain/PieceProvider.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}

module EagerWords.Server.Domain.PieceProvider (
    PieceProvider(..)
  , EagerWords.Server.Domain.PieceProvider.take
  , takePieces
  , swapOne
  , pieceProviderType
  , mkDefaultCyclicPieceProvider
  )
  where

import EagerWords.Common.Domain.Piece (Piece, Piece(Piece))
import qualified EagerWords.Common.Domain.PieceProviderType as PieceProviderType
import EagerWords.Common.Domain.PieceProviderType
import EagerWords.Server.Domain.GameError (GameError)

-- The piece generator types are closed in this implementation.
-- TODO. Would be nice to have an open piece generator implementation model.
-- Had some type system issues using type classes.

-- TODO. Name cyclic constructor parameters.

-- | Piece generator.
--   Included in the common package to allow client tests
--   to generate pieces consistently with the server.
data PieceProvider =
  RandomPieceProvider { counter :: Integer, randomizer :: IO Char} |
  CyclicPieceProvider Integer String

take :: PieceProvider -> IO (Piece, PieceProvider)

take (provider @ RandomPieceProvider {counter, randomizer}) = do
  letter <- randomizer
  let piece = Piece letter (show counter)
  let nextProvider = RandomPieceProvider (counter + 1) randomizer
  return (piece, nextProvider)

take (CyclicPieceProvider count cycler) = do
  let count' = count + 1
      piece = Piece (head cycler) (show count')
  return (piece, CyclicPieceProvider count' (drop 1 cycler))

-- TODO. Best practices to disambiguate against Prelude [take]?
take' :: PieceProvider -> IO (Piece, PieceProvider)
take' = EagerWords.Server.Domain.PieceProvider.take

takePieces :: PieceProvider -> Int -> IO ([Piece], PieceProvider)
takePieces provider max = takePiecesAux provider [] max

takePiecesAux :: PieceProvider -> [Piece] -> Int -> IO ([Piece], PieceProvider)
takePiecesAux provider list n =
  if n <= 0
    then return (list, provider)
    else do
      (piece, provider1) <- take' provider
      (pieces, provider2) <- takePiecesAux provider1 (piece:list) (n - 1)
      return (pieces, provider2)

swapOne :: PieceProvider -> Piece -> IO (Piece, PieceProvider)
swapOne provider piece = do
  (swappedPiece, provider1) <- take' provider
  -- provider2 <- give provider1 piece
  return (swappedPiece, provider1)

pieceProviderType :: PieceProvider -> PieceProviderType
pieceProviderType (RandomPieceProvider _ _) = PieceProviderType.Random
pieceProviderType (CyclicPieceProvider _ _) = PieceProviderType.Cyclic

caps = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
mkDefaultCyclicPieceProvider :: PieceProvider
mkDefaultCyclicPieceProvider = CyclicPieceProvider 0 (cycle caps)
